### ISSUE TITLE
[0.5.1] - UI issues with logo update (caching + error message)

### ISSUE THREAD

**Issue by emkll** [freedomofpress/securedrop#2844]
# Bug

Two minor UI bugs in the Instance configuration of the journalist interface.

## Steps to Reproduce
1. Logo is not immediately updated:
- Log into journalist interface
- Click on admin
- Click on instance config
- Select an image to upload
- Click on update logo
- Observe that the image is not updated immediately (in both sidebar and image selection screen). It appears to be cached by the browser.

2. Inconsistent error message:

- Log into journalist interface
- Click on admin
- Click on instance config
- Click on update logo (without selecting a file)
- Observe error message is not displayed in the flash banner (the same way a successful upload message is displayed)

## Expected Behavior

1. User should be provided immediate feedback (i.e. the image they uploaded should be visible)
2. Error message should be consistent across actions of a same page:

 
## Actual Behavior
1. Force reload is required to update the logo's on both sidebar and image selection area
2. Error message is inconsistent:
![screenshot from 2018-01-12 15-14-00](https://user-images.githubusercontent.com/15223328/34893682-5ca9dadc-f7ac-11e7-8010-7511b01ebda3.

**Comment by redshiftzero** [comment]
1 indeed sounds like a bug

On 2, there should be an error message if no file is selected displayed below the form (intended to be as close as possible to: http://uipatterns.io/input-validation/form-validation):

![screen shot 2018-01-12 at 12 32 48 pm](https://user-images.githubusercontent.com/7832803/34894160-15a6eae2-f795-11e7-9f0b-cb70da9e63af.png)

Was that error message present @emkll?

**Comment by emkll** [comment]
@redshiftzero Yes, the error message was present:
![screenshot from 2018-01-12 15-14-00](https://user-images.githubusercontent.com/15223328/34894238-8e93fa30-f7ae-11e7-8ba2-869b06b741b8.png)

**Comment by conorsch** [comment]
Can confirm that a full refresh of the new **Instance Configuration** page is necessary for the new logo to be displayed _on the Instance Configuration page_. Even without a full refresh, though, subsequent loads of the Source Interface indeed show the updated logo immediately:

![there-can-be-only-one-sd-logo](https://user-images.githubusercontent.com/657862/34895700-ba3f69b6-f79b-11e7-9f10-fb319509645e.png)

A subsequent update showed the same behavior (note this is _after_ uploading a non-kitten image):

![sdlogo-1](https://user-images.githubusercontent.com/657862/34895786-0a2be800-f79c-11e7-9a5a-9dc7980594a5.png)

As before, on the Source Interface a refresh shows it (i.e. the non-kitten image) immediately:

![sd-logo2](https://user-images.githubusercontent.com/657862/34895816-28789254-f79c-11e7-90a9-15692bfc19f1.png)

Did not see the **File required** error during normal interactive testing. When I tried clicking **Update Logo** without specifying a file, then it displayed, as a non-flashed-banner, as indicated by @emkll.

**Comment by redshiftzero** [comment]
Great, thanks for confirming @conorsch. That's definitely confusing behavior and needs to be resolved ideally without changing strings in the interface. 

Regarding 2, since the error message is displayed under the field, I think we should leave as is (happy to discuss further if anyone disagrees) for 0.5.1, though to your point @emkll I agree that UX could be better. Specifically, the flashed success message that appears at the top of the page should really appear next to the form in question, i.e. under the button (if this page were to get much bigger, we'd want to do that anyway instead of flashed messages always appearing at the top of the page). Will file a followup for this to improve this UX on the next cycle.

**Comment by conorsch** [comment]
> Regarding 2, since the error message is displayed under the field, I think we should leave as is (happy to discuss further if anyone disagrees) for 0.5.1

:+1: Let's keep the train chuggin'.

> Will file a followup for this to improve this UX on the next cycle.

The current issue seems sufficient for both tracked changes, unless you'd prefer to tackle the changes separately (thereby closing separate PRs). Either way, totally agree on documenting for now and shipping what we've got—does not strike me as a showstopper. 

Thanks for the detailed report, @emkll !

**Comment by redshiftzero** [comment]
When implementing the logo update feature, I kept the filename the same such that a single file could be whitelisted `rw` in AppArmor and updated in place by admins. The also sidesteps the issue of tracking which file should be used as the logo (which we would need to address if we included the version in the filename, thus allowing the filename to change). The downside of this approach is that the browser caches the image. I used a query parameter to bust the cache in the Jinja2 `autoversion` approach. 

In a development VM this version parameter appears:

```
<p>
  <img src="/static/i/logo.png?v=1515798259.0" class="logo small" alt="SecureDrop" width="250px">
</p>
```

However, it is absent in a staging VM:

```
<p>
  <img src="/static/i/logo.png" class="logo small" alt="SecureDrop" width="250px">
</p>
```

Digging...

**Comment by redshiftzero** [comment]
In the logic of the `autoversion` Jinja2 filter, it returns a filename without a version parameter if the file could not be located on disk by the web application code. In this case, it was because `autoversion` used a relative path, which happened to work in the development environment because the development server runs in the same directory as `static/i/logo.png`. In staging or prod however, this file could not be found and as such the file was returned without a version parameter. If we resolve this by using absolute paths, then autoversion will work as intended in the staging environment.

**Comment by redshiftzero** [comment]
Item 1 fixed in #2845, item 2 tracked in followup #2876. Closing!
